### PR TITLE
Feat/add withdrawals root

### LINKED ISSUE #2344: EIP-4895: Add `withdrawals_root` field to `Block` struct
Now that EIP-4895 is active, please add `withdrawals_root` to the `Block`. Currently have to get it from `block.other` and I envision this field will be necessary for users of `ethers-rs` soon.

### PR DESCRIPTION
**Pull request by merklefruit**
## Motivation

Adds support for [EIP-4895](https://eips.ethereum.org/EIPS/eip-4895) and closes #2344 

## Solution

Add `withdrawals_root` property to the `Block` struct type.

## PR Checklist

-   [ ] Added Tests
-   [ ] Added Documentation
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-core/src/types/block.rs [hunk 1]
```diff
@@ -84,6 +84,10 @@ pub struct Block<TX> {
     /// Base fee per unit of gas (if past London)
     #[serde(rename = "baseFeePerGas")]
     pub base_fee_per_gas: Option<U256>,
+    /// Withdrawals root hash (EIP-4895)
+    #[serde(rename = "withdrawalsRoot")]
+    #[cfg(not(feature = "celo"))]
+    pub withdrawals_root: Option<H256>,
```
Comment (mattsse) [inline ethers-core/src/types/block.rs:90]: this field is not present pre shanghai and should be `skip_serialize_if Option::is_none`

**Review by mattsse — CHANGES_REQUESTED** [review]
while we're at it, let's also add the `withdrawals` field as well

**Comment by merklefruit** [comment]
Added basic support for `Withdrawal` type and added it to the `Block` struct

File: ethers-core/src/types/withdrawal.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+/// A validator withdrawal from the consensus layer.
+/// See EIP-4895: Beacon chain push withdrawals as operations.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Withdrawal {
+    /// Monotonically increasing identifier issued by consensus layer
+    pub index: U64,
+
+    /// Index of validator associated with withdrawal
+    pub validator: U64,
```
Comment (mattsse) [inline ethers-core/src/types/withdrawal.rs:12]: add serde name to `validatorIndex`:

```json
    {
      "index": "0x1899c",
      "validatorIndex": "0x52b11",
      "address": "0xb9d7934878b5fb9610b3fe8a5e441e8fad7e293f",
      "amount": "0x51ae3c69"
    }
```

File: ethers-core/src/types/block.rs [hunk 3]
```diff
@@ -84,10 +84,14 @@ pub struct Block<TX> {
...
     #[serde(rename = "baseFeePerGas")]
     pub base_fee_per_gas: Option<U256>,
-    /// Withdrawals root hash (EIP-4895)
-    #[serde(rename = "withdrawalsRoot")]
+    /// Withdrawals root hash (if past Shanghai)
+    #[serde(skip_serializing_if = "Option::is_none", rename = "withdrawalsRoot")]
     #[cfg(not(feature = "celo"))]
     pub withdrawals_root: Option<H256>,
+    /// Withdrawals (if past Shanghai)
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (mattsse) [inline ethers-core/src/types/block.rs:92]:
```suggestion
    #[serde(default, skip_serializing_if = "Option::is_none")]
```

**Review by mattsse — CHANGES_REQUESTED** [review]
almost there

File: ethers-core/src/types/withdrawal.rs (hunk 2 above)
Comment (mattsse) [reply ethers-core/src/types/withdrawal.rs:12]: or probably better to rename the field to `validator_index` as well

Comment (merklefruit) [reply ethers-core/src/types/withdrawal.rs:13]: ok, renaming it rn

**Review by mattsse — APPROVED** [review]
lgtm

File: ethers-core/src/types/withdrawal.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+use crate::types::{Address, U256, U64};
+use serde::{Deserialize, Serialize};
+
+/// A validator withdrawal from the consensus layer.
+/// See EIP-4895: Beacon chain push withdrawals as operations.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
```
Comment (gakonst) [inline ethers-core/src/types/withdrawal.rs:6]: nit can derive

**Review by gakonst — APPROVED** [review]

**Pull request merged**